rfcs/primitive_lists_api.md

```diff
@@ -0,0 +1,283 @@
+public class PrimitiveBooleanRealmList extends PrimitiveRealmList<Boolean>
+
+protected abstract class PrimitiveRealmResults<Object> extends OrderedRealmCollection<Object>
+public class PrimitiveIntegerRealmResults extends PrimitiveRealmResults<Integer>
+public class PrimitiveFloatRealmResults extends PrimitiveRealmResults<Float>
+public class PrimitiveDoubleRealmResults extends PrimitiveRealmResults<Double>
+public class PrimitiveStringRealmResults extends PrimitiveRealmResults<String>
+public class PrimitiveBooleanRealmResults extends PrimitiveRealmResults<Boolean>
+```
+
+2) We need the same optional methods in `RealmCollection` and `OrderedRealmCollection` unless we we create an entirely
+   new collection hierarchy.
+
+3) `PrimitiveRealmQuery` will only contain the features supported by primitive arrays. If using 1b, we can even ensure
+   that all methods are useful
+
+```
+beginsWith // Only String
+between // Only numbers, date
+contains // Only String
+distinct
+endsWith // Only String
+equalTo
+findAllSorted
+findAllSortedAsync
+greaterThan // Only numbers, date
+greaterThanOrEqualTo // only numbers, date
+in
+isEmpty
+isNotEmpty
+isNotNull
+isNull
+lessThan // only numbers, date
+lessThanOrEqualTo // only numbers, date
+like // only String
+max // only numbers, date
+maximumDate // only numbers, date
+min // only numbers, date
+minimumDate // only numbers, date
+notEqualTo
+sum // only numbers // only numbers
+````
+
+# Discussion
+
+## Design 1: Relax type requirements
+
+**Advantages**
+* No need to add new types.
+* RealmQuery/RealmResults already contain methods that doesn't work in certain circumstances.
+* Schemas can already be restrict using modules, making `<E extends RealmModel>` inaccurate anyway.
+* RealmCollection will have to contain optional methods no matter what, so adding them to RealmQuery doesn't break
+  that pattern.
+* Would allow us to remove the requirement that people implements `RealmModel`, they could just use the `@RealmClass`
+  annotation.
+
+**Disadvantages**
+* API gets a lot less type safe (IntelliJ plugin might help here).
+* `RealmQuery` will suddenly have a ton of new methods that will only work for primitive arrays (lint check?)
+
+
+## Design 2: New collection classes:
+
+**Advantages**
+* Intent is more clear in the type system.
+* `RealmQuery` is not flooded with a lot of new methods.
+* No changes to existing code. Primitive arrays will be a complete new addon.
+* With 1b) we statically check everything using the type system on compile time.
+
+**Disadvantages**
+* Explosion in classes in the Realm API. Maintenance and test will be problematic.
+* It is not clear how to handle primitive arrays in the current Collection hierarchy. It seems a bit pointless to allow
+  optional methods for primitive lists in the interface if we go to great lengths in the actual class to make it type
+  safe.
+* Type safety is not guaranteed anyway if we allow `<Object>`` as generic.
+
```

Comment:
Note that, to the contrary, type safety is still guaranteed for all of RealmList.  It just isn't guaranteed here.